File: 1605_find_valid_matrix_given_row_and_column_sums/Solution.swift
// LeetCode 1605 - Find Valid Matrix Given Row and Column Sums
// https://leetcode.com/problems/find-valid-matrix-given-row-and-column-sums/

class Solution {
    func restoreMatrix(_ rowSum: [Int], _ colSum: [Int]) -> [[Int]] {
        var rowSum = rowSum
        var colSum = colSum
        var ans = [[Int]](repeating: [Int](repeating: 0, count: colSum.count), count: rowSum.count)
        var i = 0, j = 0
        while i < rowSum.count && j < colSum.count {
            let x = min(rowSum[i], colSum[j])
            ans[i][j] = x
            rowSum[i] -= x
            colSum[j] -= x
            if rowSum[i] == 0 { i += 1 }
            if colSum[j] == 0 { j += 1 }
        }
        return ans
    }
}

File: 1605_find_valid_matrix_given_row_and_column_sums/solution.rs
// LeetCode 1605 - Find Valid Matrix Given Row and Column Sums
// https://leetcode.com/problems/find-valid-matrix-given-row-and-column-sums/

impl Solution {
    pub fn restore_matrix(mut row_sum: Vec<i32>, mut col_sum: Vec<i32>) -> Vec<Vec<i32>> {
        let mut ans = vec![vec![0; col_sum.len()]; row_sum.len()];
        let (mut i, mut j) = (0usize, 0usize);
        while i < row_sum.len() && j < col_sum.len() {
            let x = row_sum[i].min(col_sum[j]);
            ans[i][j] = x;
            row_sum[i] -= x;
            col_sum[j] -= x;
            if row_sum[i] == 0 {
                i += 1;
            }
            if col_sum[j] == 0 {
                j += 1;
            }
        }
        ans
    }
}
